Synchronize RRef.to_here() CUDA Streams properly

pbelevich opened the pull request:
This PR fixes #54771

1. `c10::Event`s are stored in OwnerRRef and `OwnerRRef::recordAllStreams`, `OwnerRRef::blockAllStreams` are introduced.
2. `LazyStreamContext` is passed from `tensorpipe_agent.cpp` to `RequestCallbackImpl::processPythonRRefFetchCall` and `RequestCallbackImpl::processPythonRemoteCall` to make proper streams synchronization.
3. `CUDAEvent` was replaced with `c10::Event`, `CUDAStream` with `c10::Stream`, delete `CudaLazyStreamContext` and make `LazyStreamContext` generic, device-agnostic.

The follow up is #55757

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55985 Change MessageType values from decimals to hexadecimals for readability
* **#54932 Synchronize RRef.to_here() CUDA Streams properly**

Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

pbelevich committed:
Synchronize RRef.to_here() CUDA Streams properly

[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] Synchronize RRef.to_here() CUDA Streams properly"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] Synchronize RRef.to_here() CUDA Streams properly"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] Synchronize RRef.to_here() CUDA Streams properly"




[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -411,6 +413,7 @@ void RequestCallbackImpl::processPythonRRefFetchCall(
       Message m =
```
this `processPythonRemoteCall` function should take the ctx as an argument and insert streams to that ctx accordingly for each device used by the RRef.

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -411,6 +413,7 @@ void RequestCallbackImpl::processPythonRRefFetchCall(
       Message m =
           PythonRRefFetchRet(std::move(*result).toIValues()).toMessage();
       m.setId(messageId);
+      rref->waitAllDevices();
```
we shouldn't call `waitAllDevices` in this way, because it currently only blocks current streams, but we don't haven't set up streams using the ctx, which means that after doing this, the streams are still not synchronized.

mrshenli replied:
this is the owner RRef, and after accessing the value from this RRef, we know that the `rpc.remote` request has already recorded CUDAEvents on this RRef, so we can access those events here. Then, we need to do two things:

1. use the `ctx->getStream(device)` API to grab a stream for each device this owner RRef uses. 
2. use owner RRef's CUDAEvents to block the streams allocated in (1). 

The reason that we need to block those streams instead of current streams is because, 1) later when tensorpipe sends out response, it will use the same ctx to get streams, and we need to make sure that comm ops inserted in those streams will wait for the pending computations inserted in the `rpc.remote` request. So one solution is to use the ctx as a proxy, where the `processFetch` add streams here and tensorpipe write reuses those streams; 2) we don't want to mess up with current/default streams, unless those streams are set by us, otherwise, the internal logic might interfere with user logic in the user main thread.

mrshenli replied:
the correct ordering should be:

On processPythonRemoteCall:

1. compute value for the RRef
2. record events and store events in RRef
3. set RRef value, this will unblock 

On processPythonRRefFetchCall:

a. get the RRef value
b. when a is done, we know that 3. is done and hence 2. is done too. So that we can use the cuda events on the RRef.
c. call `ctx-getStream(device)` for each device used by the RRef, and then let RRef's CUDAEvent block the stream returned by `ctx-getStream(device)`. With this, the subsequent TensorPipe pipe->write will use the same streams and hence the ordering will be preservinged, i.e., TensorPipe comm must happen after all pending CUDA ops that contributes to the RRef value.

mrshenli commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -258,6 +258,23 @@ void OwnerRRef::setError(std::exception_ptr eptr) {
+  cudaEvents_.clear();
+  for (auto deviceIndex : deviceIndices) {
+    at::cuda::CUDAEvent cudaEvent;
+    cudaEvent.record(at::cuda::getCurrentCUDAStream(deviceIndex));
```
yep, it's OK to record on current streams, as the agent has set up current streams using the ctx.

mrshenli commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -258,6 +258,23 @@ void OwnerRRef::setError(std::exception_ptr eptr) {
+
+void OwnerRRef::waitAllDevices() {
+  for (at::cuda::CUDAEvent& cudaEvent : cudaEvents_) {
+    cudaEvent.block(at::cuda::getCurrentCUDAStream(cudaEvent.device_index()));
```
here, we shouldn't block the current streams, because this will be call in `processPythonRRefFetchCall`, and with this message, there is no CUDA arguments and hence the agent will setup current streams properly. Instead, we need to grab streams from the ctx and block those streams.

mrshenli commented on torch/csrc/distributed/rpc/request_callback.cpp:
```diff
@@ -9,13 +9,13 @@ namespace rpc {
 using namespace torch::distributed::autograd;
 
-std::shared_ptr<JitFuture> RequestCallback::operator()(Message& request) const {
+std::shared_ptr<JitFuture> RequestCallback::operator()(Message& request, const std::set<c10::DeviceIndex>& deviceIndices) const {
```
sorry that I mislead you in the simpler solution mentioned in #54771

mrshenli reviewed (changes requested):
Hey @pbelevich, apologize that the simpler approach mentioned in the #54771 won't work, and we will have to go through the solution that passes the ctx instead of the device list to `cb_->operator()(message, ctx)` and then propagate the `ctx` to both `processPythonRemoteCall` and `processPythonRRefFetchCall`.


The reason is that we will need to make sure that `processPythonRRefFetchCall` synchronizes the same
streams used by TensorPipe `pipe->write` for the fetched tensors. More specifically `processPythonRRefFetchCall` runs the following steps:

1. enter from `cb_->operator()`

https://github.com/pytorch/pytorch/blob/faa4da49ff6fdde8411572fe907e861e95ce2c85/torch/csrc/distributed/rpc/tensorpipe_agent.cpp#L801

2. get RRef value in `processPythonRRefFetchCall`

https://github.com/pytorch/pytorch/blob/faa4da49ff6fdde8411572fe907e861e95ce2c85/torch/csrc/distributed/rpc/request_callback_impl.cpp#L409

3. calls `pipeWrite` to send the responses 

https://github.com/pytorch/pytorch/blob/faa4da49ff6fdde8411572fe907e861e95ce2c85/torch/csrc/distributed/rpc/tensorpipe_agent.cpp#L715

4. When serializing responses in `tensorpipeSerialize`, grab streams from the ctx. If the ctx does not contain streams for the requested device, it will grab new streams from the stream pool. And then TensorPipe will use this streamm to communicate tensors to the caller. [IMPORTANT] So, we will need to make sure that step 2 above already created such streams in the ctx and these streams are properly synchronized using owner RRef's CUDAEvents. 

https://github.com/pytorch/pytorch/blob/faa4da49ff6fdde8411572fe907e861e95ce2c85/torch/csrc/distributed/rpc/tensorpipe_agent.cpp#L645-L646

https://github.com/pytorch/pytorch/blob/faa4da49ff6fdde8411572fe907e861e95ce2c85/torch/csrc/distributed/rpc/tensorpipe_utils.cpp#L126

pbelevich committed:
Update on "[WIP] Synchronize RRef.to_here() CUDA Streams properly"




[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/request_callback.cpp:
```diff
@@ -9,13 +9,13 @@ namespace rpc {
 using namespace torch::distributed::autograd;
 
-std::shared_ptr<JitFuture> RequestCallback::operator()(Message& request) const {
+std::shared_ptr<JitFuture> RequestCallback::operator()(Message& request, std::shared_ptr<LazyStreamContext> ctx) const {
```
nit: since we are going to access `LazyStreamContext` in general RPC files, it probably no longer belong to `tensorpipe_utils.h`. One candidate new location might be `rpc/utils.h`.

pbelevich committed:
Update on "[WIP] Synchronize RRef.to_here() CUDA Streams properly"




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] Synchronize RRef.to_here() CUDA Streams properly"




[ghstack-poisoned]

mrshenli commented on caffe2/CMakeLists.txt:
```diff
@@ -1162,6 +1162,17 @@ if(USE_DISTRIBUTED)
       USE_TENSORPIPE
     )
+    # This is needed for torch/csrc/distributed/rpc/macros.h to work
+    # in other than tensorpipe_agent source files, like rref_impl.cpp
```
in other than tensorpipe_agent source files, like rref_impl.cpp 

-> in libtorch_distributed_sources source files, e.g., rref_impl.cpp. See tools/build_variables.bzl

pbelevich replied:
I'm sorry, is it a suggestion to edit? replace the last line?

mrshenli replied:
Yep, please feel free to rephrase. I just wanna help make this comments more accurate. Because not all "other than tensorpipe_agent source files" need this. IIUC, only files not in the list below need that.

https://github.com/pytorch/pytorch/blob/b80c6f863f2327c712c478f67c248b94d66b65ac/caffe2/CMakeLists.txt#L365-L372

mrshenli commented on torch/csrc/distributed/rpc/request_callback.cpp:
```diff
@@ -9,13 +9,13 @@ namespace rpc {
 using namespace torch::distributed::autograd;
 
-std::shared_ptr<JitFuture> RequestCallback::operator()(Message& request) const {
+std::shared_ptr<JitFuture> RequestCallback::operator()(Message& request, std::shared_ptr<LazyStreamContext> ctx) const {
```
Let's break into long line into shorter pieces. Can you try run `clang-format -i` on these files.

pbelevich replied:
addressed

mrshenli commented on torch/csrc/distributed/rpc/request_callback.h:
```diff
@@ -6,13 +6,15 @@ namespace torch {
  public:
   // Invoke the callback.
-  std::shared_ptr<JitFuture> operator()(Message& request) const;
+  std::shared_ptr<JitFuture> operator()(Message& request, std::shared_ptr<LazyStreamContext> ctx) const;
```
ditto

pbelevich replied:
addressed

mrshenli commented on torch/csrc/distributed/rpc/request_callback.h:
```diff
@@ -24,7 +26,7 @@ class TORCH_API RequestCallback {
   // expected to ensure delivery of the response/exception based on their
   // implementation specific mechanisms.
-  virtual std::shared_ptr<JitFuture> processMessage(Message& request) const = 0;
+  virtual std::shared_ptr<JitFuture> processMessage(Message& request, std::shared_ptr<LazyStreamContext> ctx) const = 0;
```
ditto

pbelevich replied:
addressed

mrshenli commented on torch/csrc/distributed/rpc/request_callback_no_python.cpp:
```diff
@@ -86,7 +87,7 @@ std::shared_ptr<JitFuture> RequestCallbackNoPython::processMessage(
           }
 
-          processRpcWithErrors(*rpc, messageType, id, retFuture);
+          processRpcWithErrors(*rpc, messageType, id, retFuture, std::move(ctx));
```
> [performance-move-const-arg] warning: std::move of the const variable 'ctx' has no effect; remove std::move() or make the variable non-const

this lint failure looks real. Do we need to make this lambda mutable?

pbelevich replied:
addressed

mrshenli commented on torch/csrc/distributed/rpc/request_callback_no_python.cpp:
```diff
@@ -519,7 +525,8 @@ void RequestCallbackNoPython::processRunWithProfilingReq(
         messageId,
-        wrappedRpcResponseFuture);
+        wrappedRpcResponseFuture,
+        {});
```
this means this PR won't fix the cuda stream sync problem when the profiler is enabled? This is totally fine, we can add that in a followup PR, but let's don't forget to add it before the branch cut. 

cc @rohan-varma

mrshenli replied:
@pbelevich we can create an issue to track this so that we won't forget.

mrshenli reviewed (commented):
1st batch of comments, will submit more later.

mrshenli reviewed (commented):
another nit: let's add some description to the PR summary and also internal diff summary to explain what are the changes made in this PR, or at least mention this PR fixes #54771

mrshenli commented on torch/csrc/distributed/rpc/rref_impl.h:
```diff
@@ -395,6 +399,28 @@ class TORCH_API OwnerRRef final : public RRef {
+  void blockAllStreams(std::shared_ptr<LazyStreamContext> ctx);
+ private:
+
+ #ifdef USE_CUDA_NOT_ROCM
```
Let's continue discussing this.

> Curious, why the different class sizes could be a problem? I thought these are compile time flags, which means when CUDA is not available, it's same as this class does not have this field, no?

> The practice shows that this header is included to cpp files with CUDA and without and it produces segfaults

Sorry, I still didn't follow. Does this mean we need to propagate `USE_CUDA` properly to those files as well? Would I be correct if I assume those files are the non `tensorpipe_*` ones in `libtorch_python_distributed_sources`? E.g., the `py_rref.h` below?

https://github.com/pytorch/pytorch/blob/305abde976d232494c6b5252111ce15523511ffe/torch/csrc/distributed/rpc/py_rref.h#L3

If so, does it mean we should propagate `USE_CUDA` to those files in the following location?

https://github.com/pytorch/pytorch/blob/305abde976d232494c6b5252111ce15523511ffe/caffe2/CMakeLists.txt#L376-L385

mrshenli commented on torch/csrc/distributed/rpc/rref_impl.h:
```diff
@@ -395,6 +399,28 @@ class TORCH_API OwnerRRef final : public RRef {
+  void blockAllStreams(std::shared_ptr<LazyStreamContext> ctx);
+ private:
+
+ #ifdef USE_CUDA_NOT_ROCM
```
BTW, looks like there is an unintentional space before the #ifdef

pbelevich replied:
addressed

mrshenli commented on torch/csrc/distributed/rpc/request_callback_no_python.h:
```diff
@@ -14,7 +14,7 @@ namespace rpc {
 class TORCH_API RequestCallbackNoPython : public RequestCallback {
  public:
-  std::shared_ptr<JitFuture> processMessage(Message& request) const override;
+  std::shared_ptr<JitFuture> processMessage(Message& request, std::shared_ptr<LazyStreamContext> ctx) const override;
```
let's break this into shorter lines too

pbelevich replied:
addressed

mrshenli commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -258,6 +259,28 @@ void OwnerRRef::setError(std::exception_ptr eptr) {
   future_->setErrorIfNeeded(std::move(eptr));
 }
 
+void OwnerRRef::recordAllStreams(std::shared_ptr<LazyStreamContext> ctx) {
```
looks like this can be `const std::shared_ptr<LazyStreamContext>&` instead?

pbelevich replied:
addressed

mrshenli commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -258,6 +259,28 @@ void OwnerRRef::setError(std::exception_ptr eptr) {
+#endif
+}
+
+void OwnerRRef::blockAllStreams(std::shared_ptr<LazyStreamContext> ctx) {
```
`std::shared_ptr<LazyStreamContext>&`?

pbelevich replied:
addressed

mrshenli reviewed (commented):
LGTM! Most of my comments are minor. Will stamp when 1) unit tests are added 2) we figure out a solution on how to define `cudaEvents_;` with and without CUDA.

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"




[ghstack-poisoned]

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"


Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

[ghstack-poisoned]

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"


Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

[ghstack-poisoned]

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"


Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

[ghstack-poisoned]

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"


Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

[ghstack-poisoned]

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"


Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

[ghstack-poisoned]

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"


Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

[ghstack-poisoned]

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"


Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -798,7 +798,7 @@ void TensorPipeAgent::respond(std::shared_ptr<tensorpipe::Pipe>& pipe) {
           std::shared_ptr<JitFuture> futureResponseMessage;
           try {
-            futureResponseMessage = cb_->operator()(requestMessage);
+            futureResponseMessage = cb_->operator()(requestMessage, ctx);
```
nit: shall we add a comment here to explain why we need to pass ctx to `operator()()`?

pbelevich replied:
addressed

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -5548,3 +5563,31 @@ def _test_rref_proxy_timeout(self, rref_proxy_api):
+    def test_rref_to_here_synchronization(self):
+        dst = worker_name((self.rank + 1) % self.world_size)
+        options = self.rpc_backend_options
+        options.set_device_map(dst, {0: 0})
```
shall we add one more test for cross-device `to_here`? E.g., the forward output can resides on `cuda:1`, and `to_here()` fetch it back to `cuda:0`.

pbelevich replied:
added `test_rref_to_here_synchronization_cross_device`

mrshenli reviewed (approved):
LGTM. Stamp to unblock. Please wait for tests. Please also add an issue to track profiler support.

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"


Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

[ghstack-poisoned]

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"


Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

[ghstack-poisoned]

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"


Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

[ghstack-poisoned]

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"


Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

[ghstack-poisoned]

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"


Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

[ghstack-poisoned]

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"


Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

[ghstack-poisoned]

malfet commented on caffe2/CMakeLists.txt:
```diff
@@ -1162,6 +1162,18 @@ if(USE_DISTRIBUTED)
+    # This is needed for torch/csrc/distributed/rpc/macros.h to work
+    # in libtorch_distributed_sources source files, e.g., rref_impl.cpp.
+    # See tools/build_variables.bzl
+    if(USE_CUDA)
+      target_compile_definitions(torch_cpu PUBLIC USE_CUDA)
+    endif()
```
This is wrong. torch_cpu library should be CUDA indepentent (i.e. torch_cpu from pure CPU build and torch_cpu from CUDA build could be used as a drop replacement)

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"


Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/utils.h:
```diff
@@ -87,6 +89,105 @@ TORCH_API void populateRemoteProfiledEvents(
+    return devices;
+  }
+
+  c10::DeviceType device_type() const {
```
other part of this class uses camel instead of snake naming

pbelevich replied:
addressed

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_utils.h:
```diff
@@ -151,23 +43,52 @@ struct TensorpipeReadBuffers {
 };
 
+#ifdef USE_CUDA_NOT_ROCM
+static const auto cuda_stream_creator = [](c10::DeviceIndex index) {
```
ditto on naming convention

pbelevich replied:
addressed

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_utils.cpp:
```diff
@@ -1,10 +1,12 @@
+#include <torch/csrc/distributed/rpc/macros.h>
 #include <torch/csrc/distributed/rpc/tensorpipe_utils.h>
```
usually the header of the cpp should be the first include?

pbelevich replied:
addressed

lw replied:
The header of the cpp should come first _in its own block_ (i.e., followed by a blank line), in which case clang-format will respect that.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -605,7 +617,14 @@ void TensorPipeAgent::pipeRead(
     }
 
-    auto ctx = createLazyStreamContext();
+    auto ctx = createLazyStreamContext(
```
invocation of this occurred 4 times in `tensorpipe_*`. Does it make sense to consolidate this into a helper function and put it in `tensorpipe_utils.*`?

pbelevich replied:
addressed

mrshenli commented on torch/csrc/distributed/rpc/utils.h:
```diff
@@ -87,6 +89,105 @@ TORCH_API void populateRemoteProfiledEvents(
+
+// A general device context class for both CPU and CUDA. If CUDA is not
+// available, all CUDA-related methods will be no-ops.
+struct TORCH_API LazyStreamContext {
```
this is beautify! Thanks for consolidating!

mrshenli reviewed (approved):
LGTM! Thanks for fixing and addressing the link issue! I left some minor comments.

pbelevich committed:
Update on "Synchronize RRef.to_here() CUDA Streams properly"

This PR fixes #54771

1. `c10::Event`s are stored in OwnerRRef and `OwnerRRef::recordAllStreams`, `OwnerRRef::blockAllStreams` are introduced.
2. `LazyStreamContext` is passed from `tensorpipe_agent.cpp` to `RequestCallbackImpl::processPythonRRefFetchCall` and `RequestCallbackImpl::processPythonRemoteCall` to make proper streams synchronization.
3. `CUDAEvent` was replaced with `c10::Event`, `CUDAStream` with `c10::Stream`, delete `CudaLazyStreamContext` and make `LazyStreamContext` generic, device-agnostic.

The follow up is #55757


Differential Revision: [D27684022](https://our.internmc.facebook.com/intern/diff/D27684022)

[ghstack-poisoned]
